lib/remap-parser/src/parser.lalrpop

```diff
@@ -0,0 +1,501 @@
+use std::str::FromStr;
+use crate::ast::*;
+use crate::lex::*;
+use diagnostic::span;
+use std::collections::BTreeMap;
+use lalrpop_util::ErrorRecovery;
+use ordered_float::NotNan;
+
+grammar<'err, 'input>(input: &'input str);
+
+extern {
+    type Location = usize;
+    type Error = Error;
+
+    enum Tok<'input> {
+        "identifier" => Token::Identifier(<&'input str>),
+        "string literal" => Token::StringLiteral(<StringLiteral<&'input str>>),
+        "integer literal" => Token::IntegerLiteral(<i64>),
+        "float literal" => Token::FloatLiteral(<NotNan<f64>>),
+        "regex literal" => Token::RegexLiteral(<&'input str>),
+        "timestamp literal" => Token::TimestampLiteral(<&'input str>),
+
+        "function call" => Token::FunctionCall(<&'input str>),
+
+        "invalid token" => Token::InvalidToken(<char>),
+
+        "reserved identifier" => Token::ReservedIdentifier(<&'input str>),
+
+        LQuery => Token::LQuery,
+        RQuery => Token::RQuery,
+
+        "if" => Token::If,
+        "else" => Token::Else,
+        "null" => Token::Null,
+        "true" => Token::True,
+        "false" => Token::False,
+
+        ";" => Token::SemiColon,
+        "\n" => Token::Newline,
+        "?" => Token::Question,
+        "|" => Token::Pipe,
+        "=" => Token::Equals,
+        "," => Token::Comma,
+        "_" => Token::Underscore,
+        ":" => Token::Colon,
+        "." => Token::Dot,
+        "!" => Token::Bang,
+
+        "+" => Token::Operator("+"),
+        "*" => Token::Operator("*"),
+        "-" => Token::Operator("-"),
+        "/" => Token::Operator("/"),
+        "%" => Token::Operator("%"),
+        "<" => Token::Operator("<"),
+        "<=" => Token::Operator("<="),
+        ">" => Token::Operator(">"),
+        ">=" => Token::Operator(">="),
+        "==" => Token::Operator("=="),
+        "!=" => Token::Operator("!="),
+
+        "&&" => Token::Operator("&&"),
+        "||" => Token::Operator("||"),
+        "??" => Token::Operator("??"),
+
+        "[" => Token::LBracket,
+        "{" => Token::LBrace,
+        "(" => Token::LParen,
+
+        "]" => Token::RBracket,
+        "}" => Token::RBrace,
+        ")" => Token::RParen,
+
+        "_r" => Token::InternalTest("_r"),
+        "_e" => Token::InternalTest("_e"),
+        "_m" => Token::InternalTest("_m"),
+        "_a" => Token::InternalTest("_a"),
+        "_q" => Token::InternalTest("_q"),
+        "_c" => Token::InternalTest("_c"),
+        "_as" => Token::InternalTest("_as"),
+        "_fn" => Token::InternalTest("_fn"),
```

Comment:
These are used to test the parser using unit tests.

The reason this exists is that it allows us to only have a few public types in the parser (e.g. `pub TestCase`), which in turn significantly reduces the generated code, and thus the compilation time. (it's also just good practice not to expose internals of the parser publicly)